### PR TITLE
implement ProviderStatus, fixes #405

### LINKED ISSUE #405: Masu should stop downloading from failed providers.
## Backround

While setting up monitoring, we found that about 8-10% of celery tasks are failing because Masu is trying to download bogus CURs for both AWS and OCP.

One example:
```
[2019-04-02 12:57:21,883: WARNING/ForkPoolWorker-1] [2019-04-02 12:57:21,883] INFO in ocp_report_downloader: Looking for manifest at /var/tmp/insights_local/';"><zzzjechoi>/20190301-20190401
[2019-04-02 12:57:21,883: INFO/ForkPoolWorker-1] Looking for manifest at /var/tmp/insights_local/';"><zzzjechoi>/20190301-20190401
[2019-04-02 12:57:21,883: WARNING/ForkPoolWorker-1] [2019-04-02 12:57:21,883] ERROR in common: Unable to extract manifest data
[2019-04-02 12:57:21,883: ERROR/ForkPoolWorker-1] Unable to extract manifest data
```

## User Story

As a Service Operator,
I want Masu to track the success or failure of report downloads over time,
So that Masu can avoid downloading reports from known-bad providers.

## Acceptance Criteria

- [ ] Masu records success or failure of download attempts
- [ ] Masu stores information about repeated download failures
- [ ] When a provider has more than N failures, Masu does not attempt to download reports
- [ ] When a user changes or updates a disabled provider, Masu resumes attempts to download reports
- [ ] (optional) The Provider UI should provide a "Test Download" button that causes Masu to validate whether a download would succeed.

### PR DESCRIPTION
**Pull request by blentz**
This implements ProviderStatus tracking and checking.

If a download attempt fails, the ProviderStatus will enter a `WARNING` state. If downloads continue to fail more than the allowed number of times, the Provider will be marked as `DISABLED` and downloads will not be attempted until the ProviderStatus is reset. Resetting provider status will be implemented in Koku in a following PR.

### CODE REVIEW

File: masu/database/provider_status_accessor.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        self._provider_uuid = provider_uuid
+        self._table = self.get_base().classes.api_providerstatus
+        self._obj = self._get_db_obj_query().first()
+
+        # default to READY if there is no previous status.
+        if self._obj is None:
+            ready_status = {'provider_id': self.provider.id,
+                            'status': ProviderStatusCode.READY,
+                            'last_message': 'none',
+                            'timestamp': datetime.now(timezone.utc),
```
Comment (dccurtis) [inline masu/database/provider_status_accessor.py]: All `datetime` stuff should be using `DateAccessor`

File: masu/database/provider_status_accessor.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            return min(maximum, (2 ** (12 + interval))) + (random.randint(0, 1000) / 1000.0)
+
+        retries = self.get_retries()
+        timestamp = self.get_timestamp()
+        backoff_threshold = timestamp + timedelta(seconds=backoff(retries,
+                                                                  maximum=86400))
+        if self.get_status() == ProviderStatusCode.WARNING and \
+                retries <= ProviderStatusAccessor.MAX_RETRIES and \
+                datetime.now(timezone.utc) <= backoff_threshold:
+            return True
```
Comment (dccurtis) [inline masu/database/provider_status_accessor.py]: Need log statements for each state transition but especially when we back off from attempting an ingest for a provider

File: masu/database/provider_status_accessor.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        """Determine if the provider is waiting to retry."""
+        def backoff(interval, maximum=64):
+            """Exponential back-off."""
+            # (12 + interval) is used to select desirable values from the
+            # exponential progression.
+            return min(maximum, (2 ** (12 + interval))) + (random.randint(0, 1000) / 1000.0)
+
+        retries = self.get_retries()
+        timestamp = self.get_timestamp()
+        backoff_threshold = timestamp + timedelta(seconds=backoff(retries,
```
Comment (dccurtis) [inline masu/database/provider_status_accessor.py]: Still thinking on the date time based backoff scheme.  Thinking around how this works with our regular cadence of polling attempts and the irregular nature of AWS report updates.  I'll get back to you :)

File: masu/database/provider_status_accessor.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            error (Exception) Exception raised during processing attempts
+
+        """
+        # stay in warning state until max retries is exceeded
+        if self.get_retries() >= ProviderStatusAccessor.MAX_RETRIES:
+            self.update_status(ProviderStatusCode.DISABLED_ERROR, error)
+
+        # stay in error state until Provider is fixed.
+        # state will be reset by provider updates made via Koku API
+        if self.get_status() not in [ProviderStatusCode.DISABLED_ERROR,
```
Comment (dccurtis) [inline masu/database/provider_status_accessor.py]: As mentioned earlier, I think we need to make this type of change in a separate PR once we get commitment on API/UI changes.  We don't want to ever lock out an account with no way of unlocking.

File: masu/processor/_tasks/download.py [hunk 5]
```diff
@@ -85,15 +87,22 @@ def _get_report_files(customer_name,
...
     try:
         downloader = ReportDownloader(customer_name=customer_name,
                                       access_credential=authentication,
                                       report_source=billing_source,
                                       provider_type=provider_type,
                                       provider_id=provider_id,
                                       report_name=report_name)
-        return downloader.get_reports(number_of_months)
+        reports = downloader.get_reports(number_of_months)
     except (MasuProcessingError, MasuProviderError, ReportDownloaderError) as err:
```
Comment (dccurtis) [inline masu/processor/_tasks/download.py]: I think we need to revisit these exceptions and where they are thrown, under what conditions etc.  I tried many different things to get the this code to be hit and where your new logic would kick in.  I couldn't figure out a way.  I tried using an S3 bucket and AWS local to mess up the manifest so it fails and removing all contents in the directory.  Nothing worked.  I stayed in the `READY` state the whole time.

In fact it would be nice to have a class that could validate that a provider's backend configuration.  That would be a great status check for us but also could be used as the ingest checkpoint to decide whether or not the download should be attempted.

If we didn't do a configuration check prior to ingest; We need to make sure the backoff logic kicks in only for user configuration error situations where the user has misconfigured something.  We do a decent job of confirming some of these things when the provider is added in koku.  But once the provider exists we have no checks to confirm that those settings are still set.  Any other errors that is outside of the control of the user we should not be backing off on since it's an internal problem that we would need to fix.

File: masu/processor/orchestrator.py [hunk 6]
```diff
@@ -87,11 +88,13 @@ def prepare(self):
         """
         async_result = None
         for account in self._accounts:
-            LOG.info('Calling get_report_files with account: %s', account)
-            async_result = get_report_files.delay(**account)
-            LOG.info('Download queued - customer: %s, Task ID: %s',
-                     account.get('customer_name'),
-                     str(async_result))
+            provider_status = ProviderStatusAccessor(account.get('provider_uuid'))
+            if provider_status.is_valid() and not provider_status.is_backing_off():
```
Comment (dccurtis) [inline masu/processor/orchestrator.py]: Back to the purpose of the `WARNING` state.  I think we could accomplish the same thing with having only the retry count.

File: masu/database/provider_status_accessor.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+
+class ProviderStatusCode(enum.IntEnum):
+    """Enum class of provider status.
+
+    This mirrors the model states in koku.api.provider.models
+
+    """
+
+    NEW = 0
```
[the remark this answers is not in the document]
Comment (blentz) [reply masu/database/provider_status_accessor.py]: Use of `NEW` would be implemented in Koku during Provider creation. If we create a provider, it should start in `NEW` and transition to `READY` during initial ingestion.

File: masu/database/provider_status_accessor.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    READY = 1
+    WARNING = 33
+    DISABLED_ERROR = 98
```
[the remark this answers is not in the document]
Comment (blentz) [reply masu/database/provider_status_accessor.py]: I understand the concern - that once a provider is disabled, there is currently no way out of that state other than manually manipulating the DB or deleting the provider and re-adding, which causes data loss.

I agree with you that we don't want to stop here when the functionality is clearly going to result in a poor user experience.

We do need to move forward and make sure there is both an Admin capability and a User capability for resetting the provider state.

File: masu/database/provider_status_accessor.py (hunk 2 above)
Comment (blentz) [reply masu/database/provider_status_accessor.py]: Why? What's the concern?

File: masu/processor/_tasks/download.py (hunk 5 above)
Comment (blentz) [reply masu/processor/_tasks/download.py]: A simple way to force an error is to add a provider with a valid S3 bucket configuration, get past the initial config check, then delete the S3 bucket.

The Boto exception that's thrown will be re-raised as a ReportDownloaderError. That will hit that block.

On the OCP side, there's even fewer validation checks, which makes testing that path much easier. Just enter an incorrect report name into the provider then try to download.

File: masu/database/provider_status_accessor.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+class ProviderStatusCode(enum.IntEnum):
+    """Enum class of provider status.
+
+    This mirrors the model states in koku.api.provider.models
+
+    """
+
+    NEW = 0
+    READY = 1
+    WARNING = 33
```
[the remark this answers is not in the document]
Comment (blentz) [reply masu/database/provider_status_accessor.py]: How would eliminating the `WARNING` state facilitate disabling a broken provider?

While you're not entirely incorrect that we could branch our behavior on the retry count, that's a forest-for-the-trees kind of solution - you're missing the bigger picture on how that one state is a part of the whole collection.

It makes no sense to have `READY` and `DISABLED` be states, but then to concoct some alternative logic for the intermediate status that's in between `READY` and `DISABLED`.

Rather than picking apart every individual state to decide whether each one is "necessary" in isolation, I'd like you to try to look at the whole and decide whether we can track at least 3-5 discrete statuses without the concept of a list of states. If you can show me a solution where we don't need the ProviderStatusCode enum at all, then we can talk. Otherwise, I don't see how this is constructive - two inconsistent ways of evaluating state is not an improvement on one consistent mechanism.

File: masu/database/provider_status_accessor.py (hunk 3 above)
Comment (blentz) [reply masu/database/provider_status_accessor.py]: The reason I went with the time-based mechanism is because it allows the poller to be configured however the admin wants, but no matter how frequently the poller runs, we're enforcing the same backoff window. 

Otherwise, the rate at which we reach error states is a function of the poller frequency. This means that the shorter the poller intervals are, the more likely we see false-positive error states.

File: masu/database/provider_status_accessor.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        Return the provider status retry count.
+
+        Args:
+            None
+        Returns:
+            (int): Status retry count
+        """
+        return self._obj.retries
+
+    def is_backing_off(self):
```
[the remark this answers is not in the document]
Comment (blentz) [reply masu/database/provider_status_accessor.py]: Fair point. Do you consider that a nice-to-have or something that blocks the PR until it's changed?

**Comment by blentz** [comment]
I've pushed these changes based on feedback:
- Use `DateAccessor` instead of `datetime`
- Move non-accessor code out of ProviderStatusAccessor into its own ProviderStatus class.
- Add additional logging to show when a provider is skipped due to validity or back-off.


Here's some logs gathered from my local dev env that shows a provider going from `READY` to `WARNING` and then an additional download attempt shows the provider skipped due to being within the backing-off period.

Successful status check:
```
[2019-05-17 17:00:51,464: DEBUG/MainProcess] Task accepted: masu.celery.tasks.check_report_updates[97aafe47-50d1-4e41-8b42-6bc6b78a002a] pid:30
[2019-05-17 17:00:53,538: DEBUG/MainProcess] pidbox received method enable_events() [reply_to:None ticket:None]
[2019-05-17 17:00:55,916: WARNING/ForkPoolWorker-2] [2019-05-17 17:00:55,916] DEBUG in status: [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Status: 1 Retries: 0
[2019-05-17 17:00:55,916: DEBUG/ForkPoolWorker-2] [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Status: 1 Retries: 0
[2019-05-17 17:00:55,916: WARNING/ForkPoolWorker-2] [2019-05-17 17:00:55,916] DEBUG in status: [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Retries: 0, Timestamp: 2019-05-17 16:07:20.348774+00:00, Threshold: 2019-05-17 18:06:15.548774+00:00
[2019-05-17 17:00:55,916: DEBUG/ForkPoolWorker-2] [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Retries: 0, Timestamp: 2019-05-17 16:07:20.348774+00:00, Threshold: 2019-05-17 18:06:15.548774+00:00
[2019-05-17 17:00:55,917: WARNING/ForkPoolWorker-2] [2019-05-17 17:00:55,917] INFO in orchestrator: Getting report files for account: {'authentication': 'arn:aws:iam::111111111111:role/CostManagement', 'billing_source': 'test-bucket', 'customer_name': 'acct10001', 'provider_type': 'AWS', 'schema_name': 'acct10001', 'provider_uuid': '6e212746-484a-40cd-bba0-09a19d132d64'}
[2019-05-17 17:00:55,917: INFO/ForkPoolWorker-2] Getting report files for account: {'authentication': 'arn:aws:iam::111111111111:role/CostManagement', 'billing_source': 'test-bucket', 'customer_name': 'acct10001', 'provider_type': 'AWS', 'schema_name': 'acct10001', 'provider_uuid': '6e212746-484a-40cd-bba0-09a19d132d64'}
[2019-05-17 17:00:55,937: INFO/MainProcess] Received task: masu.processor.tasks.get_report_files[4f56b8b2-ff92-4615-ac04-8156f347f0ac]  
[2019-05-17 17:00:55,937: WARNING/ForkPoolWorker-2] [2019-05-17 17:00:55,937] INFO in orchestrator: Download queued - customer: acct10001, Task ID: 4f56b8b2-ff92-4615-ac04-8156f347f0ac
[2019-05-17 17:00:55,937: INFO/ForkPoolWorker-2] Download queued - customer: acct10001, Task ID: 4f56b8b2-ff92-4615-ac04-8156f347f0ac
```

Status changed from `READY` to `WARNING` due to an AccessDenied exception thrown by Boto:
```
[2019-05-17 17:00:56,559: ERROR/ForkPoolWorker-1] masu.processor.tasks.get_report_files[4f56b8b2-ff92-4615-ac04-8156f347f0ac]: An error occurred (AccessDenied) when calling the AssumeRole operation: Access denied
[2019-05-17 17:00:57,328: WARNING/ForkPoolWorker-1] [2019-05-17 17:00:57,328] DEBUG in status: Updating provider status: Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Status: ProviderStatusCode.WARNING, Error: An error occurred (AccessDenied) when calling the AssumeRole operation: Access denied
[2019-05-17 17:00:57,328: DEBUG/ForkPoolWorker-1] Updating provider status: Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Status: ProviderStatusCode.WARNING, Error: An error occurred (AccessDenied) when calling the AssumeRole operation: Access denied
[2019-05-17 17:00:57,337: ERROR/ForkPoolWorker-1] Task masu.processor.tasks.get_report_files[4f56b8b2-ff92-4615-ac04-8156f347f0ac] raised unexpected: ReportDownloaderError('An error occurred (AccessDenied) when calling the AssumeRole operation: Access denied',)
Traceback (most recent call last):
<... snipped ...>
masu.external.report_downloader.ReportDownloaderError: An error occurred (AccessDenied) when calling the AssumeRole operation: Access denied
```

Backing off:
```
[2019-05-17 17:03:57,139: INFO/MainProcess] Received task: masu.celery.tasks.check_report_updates[740615cd-ccd0-49a8-8866-47af710941d9]  
[2019-05-17 17:03:57,139: DEBUG/MainProcess] TaskPool: Apply <function _fast_trace_task at 0x7f37a0397488> (args:('masu.celery.tasks.check_report_updates', '740615cd-ccd0-49a8-8866-47af710941d9', {'lang': 'py', 'task': 'masu.celery.tasks.check_report_updates', 'id': '740615cd-ccd0-49a8-8866-47af710941d9', 'shadow': None, 'eta': None, 'expires': None, 'group': None, 'retries': 0, 'timelimit': [None, None], 'root_id': '740615cd-ccd0-49a8-8866-47af710941d9', 'parent_id': None, 'argsrepr': '()', 'kwargsrepr': '{}', 'origin': 'gen40@masu-7-gp8p4', 'reply_to': '2bda953f-a994-36c7-9b60-57397b879974', 'correlation_id': '740615cd-ccd0-49a8-8866-47af710941d9', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}}, '[[], {}, {"callbacks": null, "errbacks": null, "chain": null, "chord": null}]', 'application/json', 'utf-8') kwargs:{})
[2019-05-17 17:03:57,141: DEBUG/MainProcess] Task accepted: masu.celery.tasks.check_report_updates[740615cd-ccd0-49a8-8866-47af710941d9] pid:30
[2019-05-17 17:03:58,540: DEBUG/MainProcess] pidbox received method enable_events() [reply_to:None ticket:None]
[2019-05-17 17:04:00,684: WARNING/ForkPoolWorker-2] [2019-05-17 17:04:00,684] DEBUG in status: [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Status: 33 Retries: 1
[2019-05-17 17:04:00,684: DEBUG/ForkPoolWorker-2] [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Status: 33 Retries: 1
[2019-05-17 17:04:00,684: WARNING/ForkPoolWorker-2] [2019-05-17 17:04:00,684] DEBUG in status: [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Retries: 1, Timestamp: 2019-05-17 17:00:57.328613+00:00, Threshold: 2019-05-17 19:12:39.328613+00:00
[2019-05-17 17:04:00,684: DEBUG/ForkPoolWorker-2] [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Retries: 1, Timestamp: 2019-05-17 17:00:57.328613+00:00, Threshold: 2019-05-17 19:12:39.328613+00:00
[2019-05-17 17:04:00,684: WARNING/ForkPoolWorker-2] [2019-05-17 17:04:00,684] DEBUG in status: [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Status: 33 Retries: 1
[2019-05-17 17:04:00,684: DEBUG/ForkPoolWorker-2] [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Status: 33 Retries: 1
[2019-05-17 17:04:00,685: WARNING/ForkPoolWorker-2] [2019-05-17 17:04:00,685] DEBUG in status: [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Retries: 1, Timestamp: 2019-05-17 17:00:57.328613+00:00, Threshold: 2019-05-17 19:44:23.728613+00:00
[2019-05-17 17:04:00,685: DEBUG/ForkPoolWorker-2] [<class 'masu.providers.status.ProviderStatus'>] Provider: 6e212746-484a-40cd-bba0-09a19d132d64, Retries: 1, Timestamp: 2019-05-17 17:00:57.328613+00:00, Threshold: 2019-05-17 19:44:23.728613+00:00
[2019-05-17 17:04:00,685: WARNING/ForkPoolWorker-2] [2019-05-17 17:04:00,685] INFO in orchestrator: Provider skipped: 6e212746-484a-40cd-bba0-09a19d132d64 Valid: True Backing off: True
[2019-05-17 17:04:00,685: INFO/ForkPoolWorker-2] Provider skipped: 6e212746-484a-40cd-bba0-09a19d132d64 Valid: True Backing off: True
```
